### DISCUSSION TITLE
rollup example with actual zk-snark implementation

### DISCUSSION THREAD

**Discussion by Engeloid** [Consensys/gnark#938] [Q&A] [answered]
## Introduction
Hi,
I'm new to golang and web3 in general (sorry for that), but trying to create a basic zk-rollup using gnark for my project thesis for my masters degree in CS.

## Explanation
Now, I have downloaded the whole rollup example, created a few functions (seen below) based on the `rollup_test.go` file to create an operator, a transfer and then updated the state using the `UpdateState()` function, where all witnesses shall be set. 

However, I can't get the rollup example to work properly for weeks. I have tried creating an `assignment` variable (as seen in all of the documentation) and assigning the Witness values of the operator object separately but that did not work, until chatGPT helped me out to just use the `operatorA.Witnesses` directly as the circuit. 

### Issue 1
in the circuit file under the `postInit()` function, `circuit.allocateSlicesMerkleProofs()` is commented out and that's all fine I suppose. Once I try to create the witnesses, by calling `witness, err := frontend.NewWitness(&operatorA.Witnesses, ecc.BN254.ScalarField())`, I get an error in the following step, as `witness` seems to be `nil`, and therefore I can't call `witness.Public()`, as seen here:
```shell
18:47:36 INF compiling circuit
18:47:36 INF parsed circuit inputs nbPublic=2 nbSecret=46
18:47:36 INF building constraint builder nbConstraints=21695
Circuit compiled
Setup groth16 rank 1 constraint system
2023/11/28 18:47:40 Could not create witnesses: can't set fr.Element from type expr.LinearExpression
exit status 1
```
Does anybody have an idea what to do? I thought the `UpdateState()` function would fill in the witnesses correctly with appropriate variable types.

### Issue 2
Another, maybe related, issue is, once I try to compile the circuit with `circuit.allocateSlicesMerkleProofs()` being uncommented/enabled in the `circuit.go` file, I get the following error, already at the `frontend.Compile()` call:

```shell
ERR parsing circuit error="runtime error: invalid memory address or nil pointer dereference\nfrontend.parseCircuit.func2\n\tcompile.go:117\nruntime.panicmem\n\tpanic.go:261\nruntime.sigpanic\n\tsignal_windows.go:364\nutils.FromInterface\n\tconvert.go:85\nbn254.(*arithEngine).FromInterface\n\tcoeff.go:106\nr1cs.(*builder).toVariable\n\tbuilder.go:323\nr1cs.(*builder).toVariables.func1\n\tbuilder.go:333\nr1cs.(*builder).toVariables\n\tbuilder.go:340\nr1cs.(*builder).Add\n\tapi.go:42\nmimc.encryptPow5\n\tencrypt.go:150\nmimc.(*MiMC).Sum\n\tmimc.go:66\nmerkle.leafSum\n\tverify.go:69\nmerkle.(*MerkleProof).VerifyProof\n\tverify.go:92\ngnark_rollup.(*Circuit).Define\n\tcircuit.go:158\n"
2023/11/28 18:37:34 Could not compile circuit: parse circuit: runtime error: invalid memory address or nil pointer dereference
frontend.parseCircuit.func2
        compile.go:117
runtime.panicmem
        panic.go:261
runtime.sigpanic
        signal_windows.go:364
utils.FromInterface
        convert.go:85
bn254.(*arithEngine).FromInterface
        coeff.go:106
r1cs.(*builder).toVariable
        builder.go:323
r1cs.(*builder).toVariables.func1
        builder.go:333
r1cs.(*builder).toVariables
        builder.go:340
r1cs.(*builder).Add
        api.go:42
mimc.encryptPow5
        encrypt.go:150
mimc.(*MiMC).Sum
        mimc.go:66
merkle.leafSum
        verify.go:69
merkle.(*MerkleProof).VerifyProof
        verify.go:92
gnark_rollup.(*Circuit).Define
        circuit.go:158

exit status 1
```

I assumed, the `UpdateState()` function would fill in all necessary witness data and the circuit would allocate all necessary variables/slices for the circuit to compile. Am I doing something wrong?

## My Code
Following a part of my main.go:
```go
package main

import (
	"fmt"
	"log"
	"math/big"
	"os"
	. "rollup/src/gnark_rollup"
	. "rollup/src/tools"

	"github.com/consensys/gnark-crypto/ecc"
	"github.com/consensys/gnark/backend/groth16"
	"github.com/consensys/gnark/frontend"
	"github.com/consensys/gnark/frontend/cs/r1cs"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/ethclient"
	"github.com/tidwall/gjson"
)


func main() {
	
	var amount uint64 = 10
	var operatorA, privKeys = CreateOP(10)

	// get info on both parties of the transaction
	sender, err := operatorA.ReadAcc(0)
	if err != nil {
		log.Fatalf("Error reading sender account: %v", err)
	}

	receiver, err := operatorA.ReadAcc(1)
	if err != nil {
		log.Fatalf("Error reading receiver account: %v", err)
	}

	// Transactions
	firstTx := NewTransfer(amount, sender.PubKey, receiver.PubKey, sender.Nonce)
	firstTx.Sign(privKeys[0], operatorA.H)

	err = operatorA.UpdateState(firstTx, 0)
	if err != nil {
		log.Fatalf("Error updating operator state: %v", err)
	}

	r1cs, err := frontend.Compile(ecc.BN254.ScalarField(), r1cs.NewBuilder, &operatorA.Witnesses)
	if err != nil {
		log.Fatalln("Could not compile circuit:", err)
	}
	fmt.Println("Circuit compiled")

	// groth16 zkSNARK: Setup
	pk, vk, err := groth16.Setup(r1cs)
	if err != nil {
		log.Fatalln("Could not setup pk and vk:", err)
	}
	fmt.Println("Setup groth16 rank 1 constraint system")

	// create witness and proof
	witness, err := frontend.NewWitness(&operatorA.Witnesses, ecc.BN254.ScalarField())
	if err != nil {
		log.Fatalf("Could not create witnesses: %v", err)
	}
	fmt.Println("witness created: ", witness)
	publicWitness, err := witness.Public()
	if err != nil {
		log.Fatalf("Could not create public witness: %v", err)
	}
	// generate the proof
	proof, err = groth16.Prove(r1cs, pk, witness)
	if err != nil {
		log.Fatalf("Could not create proof using pk and witnesses: %v", err)
	}
	fmt.Println("Proof created")

	// verify the proof
	err = groth16.Verify(proof, vk, publicWitness)
	if err != nil {
		log.Fatalf("invalid proof: %v", err)
	}
}
```

Following the functions I copy pasted from `rollup_test.go` to create an operator:

```go
func CreateOP(nbAccounts int) (Operator, []eddsa.PrivateKey) {
	// TODO: create Rollup object and create respective smart contract
	op, userKeys := genOperator(nbAccounts)

	return op, userKeys
}

// returns a single account on the rollup
func genAccount(i int) (Account, eddsa.PrivateKey) {

	var acc Account
	var rnd fr.Element
	var privkey eddsa.PrivateKey

	// create account, the i-th account has a balance of 20+i
	acc.index = uint64(i)
	acc.Nonce = uint64(i)
	acc.balance.SetUint64(uint64(i) + 20)
	rnd.SetUint64(uint64(i))
	src := rand.NewSource(int64(i))
	r := rand.New(src)

	pkey, err := eddsa.GenerateKey(r)
	if err != nil {
		panic(err)
	}
	privkey = *pkey

	acc.PubKey = privkey.PublicKey

	return acc, privkey
}

// Returns a newly created operator and the private keys of the associated accounts
func genOperator(nbAccounts int) (Operator, []eddsa.PrivateKey) {

	operator := NewOperator(nbAccounts)

	operator.Witnesses.allocateSlicesMerkleProofs()

	userAccounts := make([]eddsa.PrivateKey, nbAccounts)

	// randomly fill the accounts
	for i := 0; i < nbAccounts; i++ {

		acc, privkey := genAccount(i)

		// fill the index map of the operator
		b := acc.PubKey.A.X.Bytes()
		operator.AccountMap[string(b[:])] = acc.index

		// fill user accounts list
		userAccounts[i] = privkey
		baccount := acc.Serialize()

		copy(operator.State[SizeAccount*i:], baccount)

		// create the list of hashes of account
		operator.H.Reset()
		operator.H.Write(acc.Serialize())
		buf := operator.H.Sum([]byte{})
		copy(operator.HashState[operator.H.Size()*i:], buf)
	}

	return operator, userAccounts

}
```

## Friendly request
I am a literal noob in all of this and would very much appreciate any Feedback in what I might be doing wrong.

**Comment by ivokub** [comment]
The given example is incomplete. I cannot get it to compile as a Go program. If you would like feedback then please provide full working example. If everything is in a file then a gist is sufficient, otherwise it should be a working git repo.

But I think it has to do that you have commented out `circuit.allocateSlicesMerkleProofs()`. There are two steps - compiling the circuit into arithmetic representation; the SNARK proofs are given based on the arithmetic representation not the Go circuit definition. Now, if you commented out `circuit.allocateSlicesMerkleProofs()`, then it means that the these parts won't be included in the arithmetization.

As for why compiling fails when you have slice allocation uncommented - dunno. Would have to see full example.

**Reply by Engeloid** [reply]
Hi, 
sorry for the late reply (although I had replied two days ago, but it somehow disappeared).
Anyway, I wanted to thank you for your feedback and explanation, and give you a more detailed view on my problem.

1.  `circuit.allocateSlicesMerkleProofs()` was commented out by the gnark developers in the first place, as seen here (latest commit: 8ed586e):
https://github.com/Consensys/gnark/blob/master/examples/rollup/circuit.go#L108
Probably because in the `rollup_test.go` file, the `allocateSlicesMerkleProofs()` function already gets called for the Witnesses before setting the MerkleProof Variables:
https://github.com/Consensys/gnark/blob/master/examples/rollup/rollup_test.go#L176

2.  I didn't find a guide or something like that for a proper way of using your examples, the rollup example in this case. So I would love to see, maybe a `main.go` example file, where we have generated all necessary rollup data like transfer amount and signature and call the `Operator.UpdateState()` function, whose Witnesses then get fed to the `frontend.Compile()` function (if I am not mistaken) and so on, until we have verified one or more transactions using zkSNARKs.

## My example
Regarding my example, here is the repo:
https://github.com/Engeloid/gnark-rollup-example/tree/discussion-branch

The `main.go` results with a `can't set fr.Element from type expr.LinearExpression` error.

**Comment by Engeloid** [comment]
Alright, I think one of the problems I have is that I try to use `OperatorA.Witnesses` for both the compilation of the circuit and for the witness creation.
To my understanding, for the circuit compilation it should be a blank circuit like so:
```
var circuit Circuit
r1cs, err := frontend.Compile(ecc.BN254.ScalarField(), r1cs.NewBuilder, &circuit)
```
And then an assignment/witnesses variable, also of type circuit, but filled with the necessary information:
```
var assignment Circuit
assignment.xy = xy
witness, err := frontend.NewWitness(&assignment, ecc.BN254.ScalarField())
```
## Current error
However, if I pass on the `&circuit` of the gnark rollup example, I get:
```
ERR parsing circuit error="runtime error: invalid memory address or nil pointer dereference\nfrontend.parseCircuit.func2\n\tcompile.go:117\nruntime.panicmem\n\tpanic.go:261\nruntime.sigpanic\n\tsignal_windows.go:364\nutils.FromInterface\n\tconvert.go:85\nbn254.(*arithEngine).FromInterface\n\tcoeff.go:106\nr1cs.(*builder).toVariable\n\tbuilder.go:323\nr1cs.(*builder).toVariables.func1\n\tbuilder.go:333\nr1cs.(*builder).toVariables\n\tbuilder.go:340\nr1cs.(*builder).Add\n\tapi.go:42\nmimc.encryptPow5\n\tencrypt.go:150\nmimc.(*MiMC).Sum\n\tmimc.go:66\nmerkle.leafSum\n\tverify.go:69\nmerkle.(*MerkleProof).VerifyProof\n\tverify.go:92\ngnark_rollup.(*Circuit).Define\n\tcircuit.go:157\n"
2023/12/11 11:42:36 Could not compile circuit: parse circuit: runtime error: invalid memory address or nil pointer dereference
frontend.parseCircuit.func2
        compile.go:117
runtime.panicmem
        panic.go:261
runtime.sigpanic
        signal_windows.go:364
utils.FromInterface
        convert.go:85
bn254.(*arithEngine).FromInterface
        coeff.go:106
r1cs.(*builder).toVariable
        builder.go:323
r1cs.(*builder).toVariables.func1
        builder.go:333
r1cs.(*builder).toVariables
        builder.go:340
r1cs.(*builder).Add
        api.go:42
mimc.encryptPow5
        encrypt.go:150
mimc.(*MiMC).Sum
        mimc.go:66
merkle.leafSum
        verify.go:69
merkle.(*MerkleProof).VerifyProof
        verify.go:92
gnark_rollup.(*Circuit).Define
        circuit.go:157

exit status 1
```

Even if I use `circuit.allocateSlicesMerkleProofs()` in the `postInit()` function.

If I leave `circuit.allocateSlicesMerkleProofs()`  commented out (as in the original circuit file), i get the following:
```
ERR parsing circuit error="runtime error: index out of range [0] with length 0\nfrontend.parseCircuit.func2\n\tcompile.go:117\nruntime.goPanicIndex\n\tpanic.go:114\nmerkle.(*MerkleProof).VerifyProof\n\tverify.go:92\ngnark_rollup.(*Circuit).Define\n\tcircuit.go:157\n"
2023/12/11 11:42:13 Could not compile circuit: parse circuit: runtime error: index out of range [0] with length 0
frontend.parseCircuit.func2
        compile.go:117
runtime.goPanicIndex
        panic.go:114
merkle.(*MerkleProof).VerifyProof
        verify.go:92
gnark_rollup.(*Circuit).Define
        circuit.go:157

exit status 1
```

## Existing Problem
So my problem is that I still can't get the rollup example to work for a simple transaction. Probably because I do not exactly know how to use the library, e.g., when to use `circuit.allocateSlicesMerkleProofs()` and I would be enlightened, if someone can show me, a working rollup example, where the circuit compiles and Witnesses can be set, such that the generated proof can be verified. 
[Link to my repo regarding this discussion post](https://github.com/Engeloid/gnark-rollup-example/tree/discuss-2)

**Reply by Engeloid** [reply, marked as answer]
After lots of trying out, I found the problem.

I wasn't allocating the slices before calling the `frontend.Define()` function. Unfortunately, it is not well documented, but I found it in this [discussion post](https://github.com/Consensys/gnark/issues/47): 

Also in the [test file](https://github.com/Consensys/gnark/blob/master/examples/rollup/circuit_test.go), if you know what you are doing, you can see in e.g. in the test in line 122 `func TestCircuitFull(t *testing.T)`, that the slices also get allocated separately there as well before `assert.ProverSucceeded()` is called.

## Solution

Allocate the slices before calling the `frontend.Define()` function. Either by making  `circuit.allocateSlicesMerkleProofs()` and also its respective variables public, i.e., changing (Uppercase) the function and variable names (e.g.: `AllocateSlicesMerkleProofs()`), such that you can call that function outside of the rollup package and allocate the slices before calling the `frontend.Define()` function.

This means you would have to allocate the slices for the circuit before the `Define()` function and for the witness naturally before assigning values to them.

## Takeaway

I know it's hard to keep the docs and inline comments up-to-date, but some things are not documented at all and I would love it if there would be put a bit more effort.
The Documentation was updated almost a year ago, and some things don't work on the spot (e.g. twistededwards has moved to `"github.com/consensys/gnark/std/algebra/native/twistededwards"` since then) and it quickly gets frustrating if tutorials that are supposed to work immediately, don't.

**Reply by ivokub** [reply]
That is good that you found a solution.

I do agree with your takeaways, namely that the documentation is not up-to-date and that we haven't really described how to work with slices (i.e. you need to pre-allocate the slices before calling `Define()`). It is actually already known problem and we're trying to resolve it mainly using two approaches:
* add more testable examples which are run (or at least compiled) to ensure that the examples are up-to-date with the current APIs and packages. For example see [proof recursion example](https://pkg.go.dev/github.com/consensys/gnark@v0.9.1/std/recursion/groth16#example-package-Native). However, we haven't found a good way to incorporate runnable examples in the documentation directly. We are thinking about migrating to [gomarkdoc](https://github.com/princjef/gomarkdoc) but it is quite lengthy process and we unfortunately do not have resources for performing the migration.
* adding functions for placeholder assignment - for example see [Groth16 recursion package](https://pkg.go.dev/github.com/consensys/gnark@v0.9.1/std/recursion/groth16#PlaceholderWitness). These placeholder init functions would automatically allocate sufficient slices so that the expected witness would fit. This seems to be working quite well. But again, we're mostly focusing on new features and updating the existing packages as needed.

In essence, designing good and usable APIs is actually quite difficult, but imo we're starting to get a good understanding and converting existing packages as we have time. And we would definitely be excited for external contributions.
